modules/promotion/commerce_promotion.module

```diff
@@ -79,3 +79,30 @@ function commerce_promotion_entity_base_field_info(EntityTypeInterface $entity_t
     return $fields;
   }
 }
+
+/**
+ * Implements hook_cron()
+ */
+function commerce_promotion_cron() {
+  commerce_promotion_disabled_expired();
+}
+
+/**
+ * Get all expired promotions that are still active and disable them.
+ */
+function commerce_promotion_disabled_expired() {
+  /** @var \Drupal\commerce_promotion\PromotionStorageInterface $promotion_storage */
+  $promotion_storage = \Drupal::service('entity_type.manager')->getStorage('commerce_promotion');
+
+  // Get query for all expired promotions.
+  $promotions = $promotion_storage->loadExpired();
+
+  if ($promotions !== FALSE) {
+    // Disable all expired promotions.
+    foreach ($promotions as $promotion) {
```

Comment:
@bojanz I think this is fine versus a queue. There shouldn't be that much activity, generally.